libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp

```diff
@@ -0,0 +1,115 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_OpticalFlow_MAV.h"
+#include <AP_Math/edc.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <utility>
+#include "OpticalFlow.h"
+#include <stdio.h>
+
+#define OPTFLOW_MAV_TIMEOUT_SEC 0.5f    // timeout after 0.5 seconds
+
+extern const AP_HAL::HAL& hal;
+
+// detect the device
+AP_OpticalFlow_MAV *AP_OpticalFlow_MAV::detect(OpticalFlow &_frontend)
+{
+    // we assume mavlink messages will be sent into this driver
+    AP_OpticalFlow_MAV *sensor = new AP_OpticalFlow_MAV(_frontend);
+    return sensor;
+}
+
+// read latest values from sensor and fill in x,y and totals.
+void AP_OpticalFlow_MAV::update(void)
+{
+    // record gyro values as long as they are being used
+    // the sanity check of dt below ensures old gyro values are not used
+    if (gyro_sum_count < 1000) {
+        const Vector3f& gyro = AP::ahrs_navekf().get_gyro();
+        gyro_sum.x += gyro.x;
+        gyro_sum.y += gyro.y;
+        gyro_sum_count++;
+    }
+
+    // return without updating state if no readings
+    if (count == 0) {
+        return;
+    }
+
+    struct OpticalFlow::OpticalFlow_state state {};
+
+    state.device_id = sensor_id;
+    state.surface_quality = quality_sum / count;
+
+    // calculate dt
+    const float dt = (latest_frame_us - prev_frame_us) * 1.0e-6;
+    prev_frame_us = latest_frame_us;
+
+    // sanity check dt
+    if (is_positive(dt) && (dt < OPTFLOW_MAV_TIMEOUT_SEC)) {
+        // calculate flow values
+        const Vector2f flowScaler = _flowScaler();
+        float flowScaleFactorX = 1.0f + 0.001f * flowScaler.x;
+        float flowScaleFactorY = 1.0f + 0.001f * flowScaler.y;
+
+        // copy flow rates to state structure
+        state.flowRate = Vector2f(((float)flow_sum.x / count) * flowScaleFactorX * dt,
+                                  ((float)flow_sum.y / count) * flowScaleFactorY * dt);
+
+        // copy average body rate to state structure
+        state.bodyRate = Vector2f(gyro_sum.x / gyro_sum_count, gyro_sum.y / gyro_sum_count);
+
+        _applyYaw(state.flowRate);
+        _applyYaw(state.bodyRate);
+    } else {
+        // first frame received in some time so cannot calculate flow values
+        state.flowRate.zero();
+        state.bodyRate.zero();
+    }
+
+    _update_frontend(state);
+
+    // reset local buffers
+    flow_sum.zero();
+    quality_sum = 0;
+    count = 0;
+
+    // reset gyro sum
+    gyro_sum.zero();
+    gyro_sum_count = 0;
```

Comment:
How can `gyro_sum_count` ever be anything but zero or one?

Reply:
the gyro_sum_count could be >1 if the update rate of the flow sensor is lower than the rate the update() is called (which is 200hz).  so for example if the sensor was 20hz but the optflow.update() is called at 200hz then there would be 10 gyro samples for every sensor sample.